File: modules/java/main/src/main/java/com/mooltiverse/oss/nyx/services/Provider.java
/*
 * Copyright 2020 Mooltiverse
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.mooltiverse.oss.nyx.services;

import com.mooltiverse.oss.nyx.services.changelog.Changelog;
import com.mooltiverse.oss.nyx.services.git.Git;
import com.mooltiverse.oss.nyx.services.github.GitHub;
import com.mooltiverse.oss.nyx.services.gitlab.GitLab;
import com.mooltiverse.oss.nyx.services.template.Template;

/**
 * These are the constants representing the available service providers.
 */
public enum Provider {
    /**
     * The changelog provider.
     */
    CHANGELOG(Changelog.class),

    /**
     * The generic Git service provider.
     */
    GIT(Git.class),

    /**
     * The <a href="https://github.com/">GitHub</a> service provider.
     */
    GITHUB(GitHub.class),

    /**
     * The <a href="https://gitlab.com/">GitLab</a> service provider.
     */
    GITLAB(GitLab.class),

    /**
     * The template service provider.
     */
    TEMPLATE(Template.class);

    /**
     * The class implementing the service
     */
    private final Class<? extends Service> serviceClass;

    /**
     * Builds the enum constant with the given implementation class.
     * 
     * @param clazz the implementation class for the service
     */
    private Provider(Class<? extends Service> clazz) {
        this.serviceClass = clazz;
    }

    /**
     * Returns the class implementing the service.
     * 
     * @return the class implementing the service.
     */
    public Class<? extends Service> getServiceClass() {
        return serviceClass;
    }
}

File: src/go/nyx/entities/provider.go
/*
 * Copyright 2020 Mooltiverse
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package entities

import (
	"fmt" // https://pkg.go.dev/fmt

	errs "github.com/mooltiverse/nyx/src/go/errors"
)

/*
These are the constants representing the available service providers.
*/
type Provider string

const (
	// The GitHub (https://github.com) service provider.
	GITHUB Provider = "GITHUB"

	// The GitLab https://gitlab.com/) service provider.
	GITLAB Provider = "GITLAB"
)

/*
Returns the string representation of the provider
*/
func (p Provider) String() string {
	switch p {
	case GITHUB:
		return "GITHUB"
	case GITLAB:
		return "GITLAB"
	default:
		// this is never reached, but in case...
		panic("unknown Provider. This means the switch/case statement needs to be updated")
	}
}

/*
Returns the service corresponding to the given string.

Errors can be:

- IllegalPropertyError in case an unknown service is passed
*/
func ValueOfProvider(s string) (Provider, error) {
	switch s {
	case "GITHUB":
		return GITHUB, nil
	case "GITLAB":
		return GITLAB, nil
	default:
		return GITHUB, &errs.IllegalPropertyError{Message: fmt.Sprintf("illegal service '%s'", s)}
	}
}
